State dictionary retrieval from offloaded modules

blbadger opened the pull request:
# What does this PR do?

Adds a function to retrieve a state dictionary from an offloaded module, loading the parameters into a specified device.  

See transformers PR [#27412](https://github.com/huggingface/transformers/pull/27412) for more information

Working on the test too:)

TODO: replace disk offload conditional check with `expand_device_map`

@SunMarc

SunMarc commented on src/accelerate/utils/modeling.py:
```diff
@@ -1536,6 +1536,52 @@ def get_state_dict_offloaded_model(model: nn.Module):
+            The name of the module of interest
+        state_dict (`Dict[str: Union(str, torch.Tensor)]`):
+            Dictionary of {module names: parameters}
+        device_to_put_offload ('cpu' or 'gpu'):
+            Device to load offloaded parameters into, defaults to the execution device if not provided.
```
```suggestion
        device_to_put_offload (Union[int, str, torch.device], *optional*):
            Device to load offloaded parameters into, defaults to the execution device if not provided.
```

SunMarc commented on src/accelerate/utils/modeling.py:
```diff
@@ -1536,6 +1536,52 @@ def get_state_dict_offloaded_model(model: nn.Module):
+
+    if preforward:
+        # in case _hf_hook is not re-initialized
+        if device_to_put_offload:
+            module._hf_hook.execution_device = original_device
+        module._hf_hook.post_forward(module, torch.tensor([]))
```
same

SunMarc reviewed (commented):
Thanks for the work @blbadger ! Let's add a test and try it with the transformers PR ! After we manage to make it work on accelerate and transformers, we can merge this! I left a few comments

blbadger commented on src/accelerate/utils/modeling.py:
```diff
@@ -1536,6 +1536,52 @@ def get_state_dict_offloaded_model(model: nn.Module):
+        # re-specify the execution device where onloaded params are sent
+        if device_to_put_offload:
+            original_device = module._hf_hook.execution_device
+            module._hf_hook.execution_device = device_to_put_offload
```
Good catch, I have modifed the `device_to_put_offload` kwarg to default to cpu

blbadger commented:
Thanks for the review @SunMarc ! I have implemented your suggestions and wrote the test, and refactored the `transformers` code to use this method. 

 I have written the execution device to switch to the cpu by default, although if you think it would be a better idea to keep this as a gpu (if available) I would be happy to change. The `transformers` test is a little more straightforward if the cpu is default, although I know that fitting code for a test i not a great design consideration:)

blbadger commented:
Just a heads up, I got around to integrating this module with transformers PR #[27412](https://github.com/huggingface/transformers/pull/27412) and testing looks good for large models there too.

SunMarc reviewed (approved):
Awesome @blbadger  ! I will merge this asap after we merge the PR that fixes the CI ! I've tested this PR with transformers and it works great !
